run/opencl/opencl_misc.h

```diff
@@ -38,6 +38,15 @@ typedef uint32_t host_size_t;
 #define NULL ((void*)0)
 #endif
 
+/* Change the below to '#if 0' if your platform can't handle __builtin_expect */
+#if 1
+#define likely(x)       __builtin_expect((x), 1)
```

Comment:
We'll need to remember to only pass a Boolean expression to `likely()`, and not say a pointer (for implied check for non-NULL). We could use `!!` in the macro to avoid this, but you'd need to retest for any performance regressions.

Reply:
FWIW, here's how we have these in `passwdqc`:

```c
#ifdef __GNUC__
#define force_inline    __attribute__ ((always_inline)) inline
#define likely(x)       __builtin_expect(!!(x), 1)
#define unlikely(x)     __builtin_expect(!!(x), 0)
#else
#define force_inline    inline
#define likely(x)       (x)
#define unlikely(x)     (x)
#endif
```

We could use similar in JtR's host code as well.

For use in OpenCL, I _guess_ we'll need:

```c
#ifdef __llvm__
#define likely(x)       __builtin_expect(!!(x), 1)
#define unlikely(x)     __builtin_expect(!!(x), 0)
#else
#define likely(x)       (x)
#define unlikely(x)     (x)
#endif
```

but this is completely untested.

Reply:
If we determine that `#ifdef __llvm__` also works, I'd prefer that. Clang 10+ sounds unnecessarily too recent, and with OpenCL this isn't exactly clang.

Reply:
The above works fine (as in positively detects it) on all my MacOS devices, as well as Linux/nvidia. I'm going with that.

Reply:
> If we determine that #ifdef __llvm__ also works

Please try that on your GTX 570 to start with. BTW I put things like `#warning "Notice: using __builtin_expect"` in there to see what's going on - but some drivers just quietly disregards that (and even `#error (...)`) while otherwise doing what they should. The only surefire way to see if we do get the built-ins is to write some actual error code in the else clause.

Reply:
`#if __llvm__` works fine on MacOS. The nvidia is busy loading hundreds of millions of hashes 🤣 but anyway I suggest we simply do it like this:
```c
#ifndef __has_builtin
#define __has_builtin(x) 0
#endif

#if __llvm__ || __has_builtin(__builtin_expect)
(...)
```

Reply:
Retested on the legacy systems, still works.